ribs/archives/_unstructured_archive.py

```diff
@@ -0,0 +1,524 @@
+            del self.__dict__["lower_bounds"]
+        if "boundaries" in self.__dict__:
+            del self.__dict__["boundaries"]
+
+        return add_info
+
+    def add_single(self, solution, objective, measures,
+                   **fields) -> Union[Any, Dict]:
+        """Inserts a single solution into the archive.
+
+        The solution is only inserted if it has a higher ``objective`` than the
+        threshold of the corresponding cell. For the default values of
+        ``learning_rate`` and ``threshold_min``, this threshold is simply the
+        objective value of the elite previously in the cell.  The threshold is
+        also updated if the solution was inserted.
+
+        .. note::
+            To make it more amenable to modifications, this method's
+            implementation is designed to be readable at the cost of
+            performance, e.g., none of its operations are modified. If you need
+            performance, we recommend using :meth:`add`.
+
+        Args:
+            solution (array-like): Parameters of the solution.
+            objective (float): Objective function evaluation of the solution.
+            measures (array-like): Coordinates in measure space of the solution.
+            fields (keyword arguments): Additional data for the solution.
+
+        Returns:
+            dict: Information describing the result of the add operation. The
+            dict contains ``status`` and ``value`` keys; refer to :meth:`add`
+            for the meaning of status and value.
+
+        Raises:
+            ValueError: The array arguments do not match their specified shapes.
+            ValueError: ``objective`` is non-finite (inf or NaN) or ``measures``
+                has non-finite values.
+        """
+        data = validate_single(
+            self,
+            {
+                "solution": solution,
+                "objective": objective,
+                "measures": measures,
+                **fields,
+            },
+        )
+
+        for name, arr in data.items():
+            data[name] = np.expand_dims(arr, axis=0)
+
+        add_info = self._store.add(
+            np.expand_dims(self.index_of_single(measures, resize=True), axis=0),
+            data,
+            {
+                "dtype": self.dtypes["threshold"],
+                "learning_rate": self._learning_rate,
+                "threshold_min": self._threshold_min,
+                "objective_sum": self._objective_sum,
+            },
+            [
+                single_entry_with_threshold,
+                compute_objective_sum,
+                compute_best_index,
+            ],
+        )
+
+        objective_sum = add_info.pop("objective_sum")
+        best_index = add_info.pop("best_index")
+
+        for name, arr in add_info.items():
+            add_info[name] = arr[0]
+
+        if add_info["status"]:
+            self._stats_update(objective_sum, best_index)
+
+        return add_info
+
+    @cached_property
```

Comment:
I used `cached_property` so that `np.max` and `np.min` didn't have to be called if the archive didn't change since they are called on the entire archive. Then the `add` method removes these properties so that the values are re-calculated any time the properties are accessed again after `add` is called.

If you're fine with keeping `cached_property` then I could make this even a little more efficient by only clearing the cache if there are `2`s in the status batch.